src/vcpkg/commands.set-installed.cpp

```diff
@@ -217,7 +243,32 @@ namespace vcpkg
 
         // currently (or once) installed specifications
         auto status_db = database_load_collapse(fs, paths.installed());
+
+        bool used_cached_compiler_info =
+            Util::any_of(action_plan.install_actions, [](const InstallPlanAction& install_action) {
+                return install_action.abi_info.value_or_exit(VCPKG_LINE_INFO)
+                    .compiler_info.value_or_exit(VCPKG_LINE_INFO)
+                    .from_cache;
+            });
+
+        if (used_cached_compiler_info)
+        {
+            if (!is_action_plan_fulfilled(action_plan, status_db))
+            {
+                // we have to install something, so get fresh compiler infos
+                for (auto& install_action : action_plan.install_actions)
+                {
+                    install_action.abi_info.clear();
+                }
+                compute_all_abis(paths, action_plan, cmake_vars, {});
+                used_cached_compiler_info = false;
+            }
+        }
         adjust_action_plan_to_status_db(action_plan, status_db);
+        if (used_cached_compiler_info)
+        {
+            Checks::check_exit(VCPKG_LINE_INFO, action_plan.empty());
+        }
```

Comment:
Now the cached compiler info is only used to detect if there is something to install, if something gets installed, fresh compiler info are used. 
(if you update your compiler the cache is not used)